### PR TITLE
Implement async_std::sync::Condvar

### PR DESCRIPTION
**Pull request by tmccombs**
Part of #217

### CODE REVIEW

**Comment by nbdd0121** [comment]
> @nbdd0121 do you have any suggestion on how to resolve your concerns?

I think notifying another waker when a WaitFuture is dropped with its opt_waker set to None should be enough. This should handle the notify_one well, but it may cause extra tasks being notified if we
have the following case:
* Task A calls wait
* Task B calls notify_all
* Task C calls wait
* Task A cancels wait

Task C will ends up being waken up. However I don't think it's a huge problem, because std's Condvar mentions that:
> Note that this function is susceptible to spurious wakeups. Condition variables normally have a boolean predicate associated with them, and the predicate must always be checked each time this function returns to protect against spurious wakeups.

**Comment by nbdd0121** [comment]
IMO we don't need `has_blocked`. Unlike Mutex which assumes most calls don't block, for Condvar I think we can assume there are tasks blocking when `notify_one` or `notify_all` is called.

**Comment by roblabla** [comment]
I feel like the wait/wait_timeout are kinda big footguns. If I understand them properly, the returned future will always return Ready the second time they're polled, without necessarily being *actually* notified. This can happen quite easily by using e.g. the `join` adapter from futures-rs.

This is *technically* okay according to Condvars' spurious wakeup conditions, but that is not mentioned at all in the docs. Besides, it seems very trivial to cause a spurious wakeup, and will likely produce bad, hard to debug code. I kinda question the usefulness of those functions compared to `wait_until`.

**Comment by tmccombs** [comment]
>  If I understand them properly, the returned future will always return Ready the second time they're polled

Kind of. They will return `Ready` if they are polled and the attached Mutex is available. I could potentially use an `Arc<AtomicBool>` to keep track of if it was _actually_ notified. But I think (though I don't have benchmarks) that in most cases that would hurt performance.

**Comment by nbdd0121** [comment]
> Kind of. They will return `Ready` if they are polled and the attached Mutex is available. I could potentially use an `Arc<AtomicBool>` to keep track of if it was _actually_ notified. But I think (though I don't have benchmarks) that in most cases that would hurt performance.

Probably only need to check if `Option<Waker>` is turned to `None`. Otherwise it'll be spurious wake up.

**Comment by tmccombs** [comment]
>  Probably only need to check if Option<Waker> is turned to None. Otherwise it'll be spurious wake up.

Yes, but that requires locking the mutex for the wakers when polled, which is doable, but could hurt performance if the condition variable has very much contention. And if `wait_until`/`wait_timeout_until` are used, then it isn't an issue.

**Comment by nbdd0121** [comment]
You need to lock the mutex to remove the waker when cancelling/completing the notify call, or to insert the waker when beginning a new call. I don't see why it'll hurt performance compared to wait_until.

**Comment by tmccombs** [comment]
Ok, I'll take a stab at implementing that. It looks like there is also now a `WakerSet` type that I should use instead of a Slab of Wakers.

**Comment by tmccombs** [comment]
@nbdd0121 I think my latest push should address your concerns.

File: src/sync/waker_set.rs [hunk 1]
```diff
@@ -80,6 +80,28 @@ impl WakerSet {
         }
     }
 
+    /// If the waker for this key is still waiting for a notification, then update
+    /// the waker for the entry, and return false. If the waker has been notified,
+    /// treat the entry as completed and return true.
+    #[cfg(feature = "unstable")]
+    pub fn remove_if_notified(&self, key: usize, cx: &Context<'_>) -> bool {
```
Comment (tmccombs) [inline src/sync/waker_set.rs]: I can't think of a really good name for this.

**Comment by nbdd0121** [comment]
WakerSet's notify_one does not wake up anything if some task is woken already. This is slightly different from Condvar's notify_one which should always wake up one task (in the case of two concurrent notify_one call for example).

**Comment by tmccombs** [comment]
If i understand the code on master that is how `notify_any` behaves, but `notify_one` behaves as we want for condvar.

**Comment by nbdd0121** [comment]
Sorry, you're right. That change is introduced in e9edadf which isn't part of a PR so I am not aware of.

File: src/sync/condvar.rs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+    /// # }) }
+    /// ```
+    pub fn notify_all(&self) {
+        self.wakers.notify_all();
+    }
+}
+
+impl fmt::Debug for Condvar {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        //f.debug_struct("Condvar").finish()
```
Comment (dignifiedquire) [inline src/sync/condvar.rs:364]: what about this comment?

File: src/sync/condvar.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        f.pad("Condvar { .. }")
+    }
+}
+
+struct AwaitNotify<'a, 'b, T> {
```
Comment (dignifiedquire) [inline src/sync/condvar.rs:369]: Could you add a description of the purpose of the struct and its fields, that will make it easiert to understand how this is structured.

File: src/sync/waker_set.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/sync/waker_set.rs:87]: seems descriptive at least :)

**Comment by tmccombs** [comment]
I rebased and addressed your comments. Should I squash as well?

**Comment by dignifiedquire** [comment]
Thank you

**Pull request merged**